### ISSUE TITLE
feat: Add tagging of feature-toggles

### ISSUE THREAD

**Issue by goober** [Unleash/unleash#655]
When you are sharing an unleash instance with several teams it can be quite cumbersome to find what team is responsible for each feature toggle. Another case is that you want to group several feature-toggles together for a specific project or a release or whatever. 

A simple solution for this would be to add the possibility to tag each feature with an arbitrary number of tags that could be used for filtering the list of feature-toggles by a given tag.

Is this a feature that you have been considered or if it is something that you think would be a good fit for adding to unleash?

**Comment by ivarconr** [comment]
Yes, this is something we want. I am also wondering if tags should have a type. This would make it easier to build integrations, e.g. connect a feature toggle to a specific Jira issue and have a webhook posting updates to the specific Jira issue.

**Comment by goober** [comment]
That is some really good ideas where tags could be used.
Can you elaborate a bit on what you mean by typed tags?

**Comment by ivarconr** [comment]
So in most products having tags they tend to just be a list of "string fields", which often is good enough. By also classifying the tag with a type we can easier use that to form decisions, when a tag changes e.g.:

```js
{
  name: 'someToggle`
  enabled: true
  tags: [
    {value: 'some-group', type: 'simple'}
    {value: 'JIRA-123', type: 'jira'},
    {value: '#some-room', type: 'slack'},
  ]
  ...
}
```

Having this extra level of tag types, could allow event listeners to pick correct data from the tags in order to send updates to the correct location in other sevices. In the example above we could have two different actions performed when there is a change in the feature toggle:

1. Post update to the jira issue: JIRA-123
2. Post update in slack channel '#some-room'

I'm just thinking out load here. There might be more things to consider when it is implemented. Types should of course be highly customizable.

**Comment by goober** [comment]
I get your point here.

As a workaround for this functionality in products where only strings are allowed is usually to prefix the tag with your type.
e.g.
```json
{
  "name": "someToggle",
  "enabled": true,
  "tags": [
    "simple:some-group",
    "jira:JIRA-123",
    "slack:#some-room"
  ]
  ...
}```

That said, I like the idea of splitting them up as in your example, but it can be achieved with simple strings as well.

**Comment by ivarconr** [comment]
Good point, thanks for the reminder. I think I need to way the pros/cons.

**Comment by chriswk** [comment]
Going to start looking at this today.
First thought is to keep the typed tags structure, allowing plugins to register new types of tags.

As in, the core feature will just support `{ value: 'some-value', type: 'simple' }` and then a slack plugin can register the `slack` type and a jira plugin can register the `jira` type. 

That way, we can filter event notification to plugins based upon what tags they've registered an interest in or tag types they've registered.

**Comment by chriswk** [comment]
The `simple` tag type would then mostly be a filter and categorization aid for the frontend.

**Comment by SimenB** [comment]
Are there any plans to expose this data to clients? From what I can see it's not included in `/api/client/features`.

My use case is we wanna include _some_ toggles in a sort of first paint data dump in the initial markup served to users. Since this of course messes up metrics horribly, it'd be great to be able to only include toggles with a given tag. Can of course be worked around via naming conventions etc, but it seems to me like a tag would solve the issue a bit more generally?

**Comment by chriswk** [comment]
> Are there any plans to expose this data to clients? From what I can see it's not included in /api/client/features.

There's the possibility of filtering on tags https://github.com/Unleash/unleash/blob/master/docs/api/admin/feature-toggles-api.md#filter-feature-toggles, which I realise isn't exposed in the sdks for now. At the moment we're not exposing tags per feature; we haven't found a good way of joining the tables for quick display when fetching all.
